middleware/CompressionHandler.md

```diff
@@ -1,47 +1,64 @@
 # Compression Handler
 
-A middleware component that requests, detects and decompresses response bodies.
+A middleware component that compresses request bodies and falls back to an uncompressed body if the API doesn't support request body compression.
 
 ## Requirements
 
-- Add `Accept-encoding: gzip` to request headers.
-- Decompress response bodies that have the header `Content-Encoding: gzip`
+- Add `Content-encoding: gzip` to request headers for requests with a body.
+- Compresses request bodies.
+- Should the handler receive a [415 Unsupported Media Type](https://httpwg.org/specs/rfc7231.html#status.415) response or a `400 Bad Request` response with an OData error containing the following message `Unable to read JSON request payload. Please ensure Content-Type header is set and payload is of valid JSON format.` when the content-type was `application/json`, the handler will retry sending the request without compressing it a maximum of one time before passing on the error.
 - Compression handler should be installed as a default handler by [GraphClientFactory](../GraphClientFactory.md)
```

Comment:
Thinking this may lead to retries a customer is unaware about and that they should make the decision to add the compression handler?

Reply:
I'm not sure I'm following the question here?

Reply:
If the compression handler is enabled by default yet it seems that not all endpoints would support compression initially, then there would be unexpected 415/400s leading to retries.
Thinking that since there's an overhead of a retry that a customer has no idea about that the they should add the handler themselves instead of making it a default.

Reply:
thanks for clarifying, as far as I understand, all the endpoints under the same host will support it or won't. So we might end-up in a situation where graph.microsoft.com does support it, but graph.microsoft.us doesn't (from an environment perspective). We also might end-up in a situation where graph.microsoft.com supports it, but outlook.office.com doesn't (large file uploads).
And yes the fact they are returning a 400 instead of a 415 today makes the retry logic tricky, it'd be much simpler if they did since the logic would be more specific.
This is why I'm suggesting retrying only `if (415 || 400 && request.ContentType.StartsWith("application/json") && response.message.Contains("Unable to read JSON request payload. Please ensure Content-Type header is set and payload is of valid JSON format."))`

Reply:
Sounds good if all endpoints under the same host behave the same way.
Also realising that there's nothing "wrong" with retrying by default since we support the Retry Handler by default as well.